[WIP] Concurrent merging with join set algorithm

zhaih opened the pull request:
### Description
In #14331  @mayya-sharipova made a great change to introduce this smart algorithm of computing join set and then use it to accelerate the graph merge. This PR tries to make the above algorithm work with the concurrent graph merge.

### Status
Reworked on how the concurrent merge works, basically now the main thread is computing the join set and distribute the `addNode` work to worker threads, while the worker thread only do one thing -- add the node to the graph. 

The PR is still **not ready for review** yet (quite close!), need to add more unit tests and also need some more refactoring & documentation

### Benchmarks
The benchmark looks great, I ran using 4 threads and 8 threads, in most cases **candidate is only using 60%-70% of the time to do the final force merge** (except one case which is 4 threads 5M docs, I don't know why but probably noise?) Notice that we **sacrifice roughly 3% of recall** in exchange. Here are the result:

#### candidate
```
4 merging threads - 4 worker
recall  latency(ms)  netCPU  avgCpuCount      nDoc  topK  fanout  maxConn  beamWidth  quantized  visited  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.793        0.241   0.236        0.979   1000000   100      50       64        250         no     1521     30.07      33252.42           43.54             1          436.65       381.470      381.470       HNSW
 0.771        0.263   0.258        0.981   2000000   100      50       64        250         no     1561     60.74      32926.15          103.67             1          877.18       762.939      762.939       HNSW
 0.749        0.279   0.274        0.982   5000000   100      50       64        250         no     1641    170.92      29253.79          347.97             1         2223.36      1907.349     1907.349       HNSW
 0.726        0.391   0.299        0.765  10000000   100      50       64        250         no     1693    379.66      26339.50          680.14             1         4477.64      3814.697     3814.697       HNSW

8 merging threads - 8 worker
 recall  latency(ms)  netCPU  avgCpuCount      nDoc  topK  fanout  maxConn  beamWidth  quantized  visited  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.791        0.245   0.240        0.980   1000000   100      50       64        250         no     1537     32.13      31121.62           27.91             1          436.62       381.470      381.470       HNSW
 0.769        0.267   0.262        0.981   2000000   100      50       64        250         no     1589     60.85      32869.33           71.18             1          876.73       762.939      762.939       HNSW
 0.749        0.276   0.271        0.982   5000000   100      50       64        250         no     1650    178.00      28090.05          174.93             1         2224.16      1907.349     1907.349       HNSW
 0.728        0.307   0.302        0.984  10000000   100      50       64        250         no     1694    395.79      25265.80          453.51             1         4475.51      3814.697     3814.697       HNSW
```

#### baseline
```
4 merging threads - 4 worker
recall  latency(ms)  netCPU  avgCpuCount      nDoc  topK  fanout  maxConn  beamWidth  quantized  visited  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.820        0.263   0.258        0.981   1000000   100      50       64        250         no     1722     28.61      34958.92           76.49             1          444.56       381.470      381.470       HNSW
 0.794        0.289   0.284        0.983   2000000   100      50       64        250         no     1753     61.73      32400.21          188.70             1          893.63       762.939      762.939       HNSW
 0.776        0.313   0.308        0.984   5000000   100      50       64        250         no     1844    183.35      27270.84          330.30             1         2272.53      1907.349     1907.349       HNSW
 0.760        0.333   0.328        0.985  10000000   100      50       64        250         no     1893    431.97      23149.70         1328.51             1         4590.19      3814.697     3814.697       HNSW

8 merging threads - 8 worker
 recall  latency(ms)  netCPU  avgCpuCount      nDoc  topK  fanout  maxConn  beamWidth  quantized  visited  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.818        0.271   0.266        0.982   1000000   100      50       64        250         no     1726     32.70      30580.10           40.93             1          444.54       381.470      381.470       HNSW
 0.802        0.287   0.282        0.983   2000000   100      50       64        250         no     1759     65.81      30389.13          103.17             1          893.75       762.939      762.939       HNSW
 0.782        0.321   0.315        0.981   5000000   100      50       64        250         no     1857    197.09      25369.38          196.99             1         2273.53      1907.349     1907.349       HNSW
 0.755        0.336   0.330        0.982  10000000   100      50       64        250         no     1918    448.10      22316.70          698.71             1         4588.53      3814.697     3814.697       HNSW
```

### Some details
One tricky thing is that there're several places (due to the nature of the join set algorithm) I have to explicitly put a "barrier" to ensure all nodes are processed before going to the next phase, so I cannot really use the very same code for both single thread and multi-thread, and also I modified the algorithm a bit so that it does not depend on previous node being added to the graph when adding the rest of nodes that are outside of join set.

## Below are old progresses
### Progress
I made the first version which just naively let each thread deal with one segment and then handle rest of the nodes like before and get a quite mixed and ineresting benchmark result. Basically the changed code only performs better in 10M doc situation, which kind of making sense since I made it handle one segment per thread, so if there's not enough segments it won't be able to make use of all the threads.

### Benchmark
#### candidate
```
recall  latency(ms)  netCPU  avgCpuCount      nDoc  topK  fanout  maxConn  beamWidth  quantized  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.784        0.254   0.249        0.980   1000000   100      50       64        250         no     28.48      35108.66           98.94             1          435.66       381.470      381.470       HNSW
 0.761        0.266   0.261        0.981   2000000   100      50       64        250         no     61.42      32562.15          231.87             1          875.69       762.939      762.939       HNSW
 0.738        0.281   0.276        0.982   5000000   100      50       64        250         no    182.75      27359.33          624.20             1         2213.94      1907.349     1907.349       HNSW
 0.716        0.310   0.295        0.952  10000000   100      50       64        250         no    394.47      25350.47          782.60             1         4456.40      3814.697     3814.697       HNSW
```

#### baseline
```
recall  latency(ms)  netCPU  avgCpuCount      nDoc  topK  fanout  maxConn  beamWidth  quantized  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.817        0.284   0.270        0.951   1000000   100      50       64        250         no     28.31      35318.22           66.91             1          444.55       381.470      381.470       HNSW
 0.794        0.303   0.288        0.950   2000000   100      50       64        250         no     63.60      31447.53          158.45             1          894.32       762.939      762.939       HNSW
 0.774        0.306   0.301        0.984   5000000   100      50       64        250         no    180.95      27632.40          253.59             1         2274.45      1907.349     1907.349       HNSW
 0.755        0.334   0.329        0.985  10000000   100      50       64        250         no    433.43      23071.94         1022.34             1         4591.37      3814.697     3814.697       HNSW
```

zhaih committed:
Pure refactoring

zhaih committed:
Finish concurrent change

msokolov commented:
the results are confusing indeed! I would suggest multiple runs with a smaller index to see if there is substantial noise?  Also - why do you say there is only a single segment in these smaller indexes?  There should be many segments produced during the indexing run.  I do see the result indicates 1 segment, but probably that is because you chose to `-forceMerge` and it just says the number of segments at the end?

zhaih committed:
Reworked, benchmarked, still need to polish and add more test cases

zhaih commented:
@msokolov 
> Also "I modified the algorithm a bit so that it does not depend on previous node being added to the graph when adding the rest of nodes that are outside of join set." makes me wonder if this might be part of the problem with recall? I'm not sure exactly what you meant (haven't reviewed the code), but it sounds as if we don't wait for all of the "covering set" to be indexed before beginning to add the remainder of the nodes?

Let me explain a bit more :)
So we do ensure the "covering set" to be fully indexed before the other part to be indexed, the one I mentioned is in [this line](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/hnsw/MergingHnswGraphBuilder.java#L168)
which says
```
        // if u's neighbour v is in the join set, or already added to gL (v < u),
        // then we add v's neighbours from gL to the candidate list
        if (v < u || j.contains(v)) {
```
I removed the `if (v < u)` part so that we don't really add anything that are "supposed" to be added previously (because that won't be the case in concurrent setting)

This are actually making the recall better, because the old benchmark (candidate) result is a more strict replication of the join set algorithm, which is showing even worse recall..

zhaih commented:
I tried to play with the beamWidth and realized the recall is actually quite hard to improve:
```
### Candidate
recall  latency(ms)  netCPU  avgCpuCount     nDoc  topK  fanout  maxConn  beamWidth  quantized  visited  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.700        0.272   0.264        0.971  5000000   100      50       48        128         no     1405     99.00      50506.58          113.60             1         2165.19      1907.349     1907.349       HNSW
 0.750        0.276   0.271        0.982  5000000   100      50       48        256         no     1643    182.84      27346.76          165.61             1         2224.54      1907.349     1907.349       HNSW
 0.771        0.294   0.290        0.986  5000000   100      50       48        512         no     1744    338.41      14774.93          239.39             1         2247.09      1907.349     1907.349       HNSW
 0.779        0.294   0.289        0.983  5000000   100      50       48        768         no     1794    487.29      10260.75          261.00             1         2257.86      1907.349     1907.349       HNSW

### Baseline
recall  latency(ms)  netCPU  avgCpuCount     nDoc  topK  fanout  maxConn  beamWidth  quantized  visited  index(s)  index_docs/s  force_merge(s)  num_segments  index_size(MB)  vec_disk(MB)  vec_RAM(MB)  indexType
 0.742        0.277   0.272        0.982  5000000   100      50       48        128         no     1604    107.16      46658.33          104.35             1         2211.86      1907.349     1907.349       HNSW
 0.782        0.306   0.301        0.984  5000000   100      50       48        256         no     1852    207.64      24080.49          193.89             1         2273.83      1907.349     1907.349       HNSW
 0.808        0.343   0.338        0.985  5000000   100      50       48        512         no     2123    375.13      13328.82          380.45             1         2336.24      1907.349     1907.349       HNSW
 0.825        0.362   0.357        0.986  5000000   100      50       48        768         no     2294    582.56       8582.85          622.42             1         2372.60      1907.349     1907.349       HNSW
```

If we look at the 0.779 recall in candidate vs. 0.782 in baseline I think the baseline definitely wins.. So for now I'll just close this PR. Tho I think some ideas here might be able to improve this overall performance so someone else/me probably can take a look in the future and grab some ideas.
